daemon/container_operations_unix.go

```diff
@@ -120,19 +122,21 @@ func (daemon *Daemon) setupIpcDirs(c *container.Container) error {
 				return err
 			}
 
-			shmSize := int64(daemon.configStore.ShmSize)
-			if c.HostConfig.ShmSize != 0 {
-				shmSize = c.HostConfig.ShmSize
-			}
-			shmproperty := "mode=1777,size=" + strconv.FormatInt(shmSize, 10)
+			shmproperty := "mode=1777,size=" + strconv.FormatInt(c.HostConfig.ShmSize, 10)
 			if err := unix.Mount("shm", shmPath, "tmpfs", uintptr(unix.MS_NOEXEC|unix.MS_NOSUID|unix.MS_NODEV), label.FormatMountLabel(shmproperty, c.GetMountLabel())); err != nil {
 				return fmt.Errorf("mounting shm tmpfs: %s", err)
 			}
 			if err := os.Chown(shmPath, rootIDs.UID, rootIDs.GID); err != nil {
 				return err
 			}
+			c.ShmPath = shmPath
 		}
 
+	case ipcMode.IsPrivate():
+		c.ShmPath = ""
+
+	default:
```

Comment:
Oh, apologies, I meant the `default:` case, because this PR invalidates what was previously valid, causing upgrades of docker to fail. Here's to illustrate what I meant:

Build from master, then:

```bash
$ docker run -dit --name foo busybox
$ docker inspect -f '{{json .HostConfig.IpcMode}}' foo
""

$ docker stop foo
```

Stop the daemon, build from your PR and start the new daemon

```bash
$ docker start foo
Error response from daemon: linux spec namespaces: Invalid IPC mode:
Error: failed to start containers: foo
```

The `foo` container was created with the "old" (master) version of Docker, but after upgrading to the new (this pr) version of docker, the container can no longer be started